File: backend_flask/swn_char/swn_char_data.py
#  SWN Char Gen © 2025. by Tod Casasent is licensed under CC BY-NC-SA 4.0. To view a copy of this license, visit https://creativecommons.org/licenses/by-nc-sa/4.0/
# 
#  Rules and values associated with the SWN system are the copyright of Kevin Crawford, Sine Nominee Publishing.
#  https://sine-nomine-publishing.myshopify.com/
#  Used in accordance with his Discord message of limiting content to the free version of the rules.
#  Specifically, the SWNSystem.java and swn_system.py game strings, values, and relationship are the copyright of Kevin Crawford, Sine Nominee Publishing.

"""
Character data module for the Stars Without Number Character Generator.

This module defines the SWNCharData dataclass, which stores the data for a
character in the Stars Without Number role-playing game system.
"""

from dataclasses import dataclass

from swn_char.swn_system import AttributeEnum


@dataclass
class SWNCharData:
	"""
	Dataclass for storing Stars Without Number character data.
	
	This class stores all the data for a Stars Without Number character,
	including character details, attributes, modifiers, and tracking for
	changed attributes.
	
	Attributes:
		mName: The character's name.
		mStrength: The character's Strength attribute.
		mDexterity: The character's Dexterity attribute.
		mConstitution: The character's Constitution attribute.
		mIntelligence: The character's Intelligence attribute.
		mWisdom: The character's Wisdom attribute.
		mCharisma: The character's Charisma attribute.
		mStrengthModifier: The modifier for the Strength attribute.
		mDexterityModifier: The modifier for the Dexterity attribute.
		mConstitutionModifier: The modifier for the Constitution attribute.
		mIntelligenceModifier: The modifier for the Intelligence attribute.
		mWisdomModifier: The modifier for the Wisdom attribute.
		mCharismaModifier: The modifier for the Charisma attribute.
		mChangedAttribute: The attribute that has been changed to a special value.
		mChangedAttributeOriginalValue: The original value of the changed attribute.
	"""
	# ####
	# Details section
	# ####
	mName: str = "Default Name"

	# ####
	# Attributes section
	# ####
	# Class variable declarations for the six attributes
	mStrength: int = 0
	mDexterity: int = 0
	mConstitution: int = 0
	mIntelligence: int = 0
	mWisdom: int = 0
	mCharisma: int = 0

	# Class variable declarations for the modifiers
	mStrengthModifier: int = 0
	mDexterityModifier: int = 0
	mConstitutionModifier: int = 0
	mIntelligenceModifier: int = 0
	mWisdomModifier: int = 0
	mCharismaModifier: int = 0

	# class variable to track which attribute is set to 14
	mChangedAttribute: AttributeEnum = AttributeEnum.NONE
	# variable to store original value of attribute set to 14
	mChangedAttributeOriginalValue: int = 0

File: backend_spring/src/main/java/org/kuroneko/swn_char_gen/backend_spring/models/SWNCharData.java
/*
 * SWN Char Gen © 2025. by Tod Casasent is licensed under CC BY-NC-SA 4.0. To view a copy of this license, visit https://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * Rules and values associated with the SWN system are the copyright of Kevin Crawford, Sine Nominee Publishing.
 * https://sine-nomine-publishing.myshopify.com/
 * Used in accordance with his Discord message of limiting content to the free version of the rules.
 * Specifically, the SWNSystem.java and swn_system.py game strings, values, and relationship are the copyright of Kevin Crawford, Sine Nominee Publishing.
 */

package org.kuroneko.swn_char_gen.backend_spring.models;

/**
 * Data class for SWN character data.
 * <p>
 * This is an echo of the Python dataclass swn_char_data. The Python class is used to get
 * "clean" data-only using to_dict. This way the methods attached to the parent swn_char class
 * can change, without the dict (used for export and import) being affected.
 * <p>
 * Java doesn't really have a setup for that, so I'm using this for now by using default (package) access.
 */
public class SWNCharData
{
	// ####
	// Details section
	// ####
	/**
	 * The character's name. Defaults to "Default Name".
	 */
	String mName = "Default Name";

	// ####
	// Attributes section
	// ####
	/**
	 * The character's Strength attribute value. Defaults to 0.
	 */
	int mStrength = 0;
	
	/**
	 * The character's Dexterity attribute value. Defaults to 0.
	 */
	int mDexterity = 0;
	
	/**
	 * The character's Constitution attribute value. Defaults to 0.
	 */
	int mConstitution = 0;
	
	/**
	 * The character's Intelligence attribute value. Defaults to 0.
	 */
	int mIntelligence = 0;
	
	/**
	 * The character's Wisdom attribute value. Defaults to 0.
	 */
	int mWisdom = 0;
	
	/**
	 * The character's Charisma attribute value. Defaults to 0.
	 */
	int mCharisma = 0;

	/**
	 * The modifier for the character's Strength attribute. Calculated based on the Strength value.
	 */
	int mStrengthModifier = 0;
	
	/**
	 * The modifier for the character's Dexterity attribute. Calculated based on the Dexterity value.
	 */
	int mDexterityModifier = 0;
	
	/**
	 * The modifier for the character's Constitution attribute. Calculated based on the Constitution value.
	 */
	int mConstitutionModifier = 0;
	
	/**
	 * The modifier for the character's Intelligence attribute. Calculated based on the Intelligence value.
	 */
	int mIntelligenceModifier = 0;
	
	/**
	 * The modifier for the character's Wisdom attribute. Calculated based on the Wisdom value.
	 */
	int mWisdomModifier = 0;
	
	/**
	 * The modifier for the character's Charisma attribute. Calculated based on the Charisma value.
	 */
	int mCharismaModifier = 0;

	/**
	 * Tracks which attribute has been changed to a special value (14).
	 * NONE indicates no attribute has been changed.
	 */
	AttributeEnum mChangedAttribute = AttributeEnum.NONE;
	
	/**
	 * Stores the original value of the attribute that was changed to 14.
	 * Used to restore the attribute if a different attribute is changed.
	 */
	int mChangedAttributeOriginalValue = 0;

}